spatialos-sdk/src/worker/schema.rs

```diff
@@ -0,0 +1,662 @@
+// the purpose of FFI.
+impl ArrayField for EntityId {
+    fn add_field_list(object: &mut SchemaObject, field: FieldId, data: &[Self::RustType]) {
+        unsafe {
+            Schema_AddEntityIdList(
+                object.internal,
+                field,
+                data.as_ptr() as *const _,
+                data.len() as u32,
+            );
+        }
+    }
+
+    fn get_field_list(object: &SchemaObject, field: FieldId, data: &mut Vec<Self::RustType>) {
+        let count = Self::field_count(object, field) as usize;
+
+        // Ensure that there is enough capacity for the elements in the schema field.
+        if data.capacity() < count {
+            data.reserve(count - data.capacity());
+        }
+
+        // Replace the contents of `data` with the list of values in the schema field.
+        unsafe {
+            data.set_len(count);
+            Schema_GetEntityIdList(object.internal, field, data.as_mut_ptr() as *mut _);
+        }
+    }
+}
+
+impl Sealed for bool {}
+
+impl SchemaField for bool {
+    type RustType = Self;
+
+    fn add_field(object: &mut SchemaObject, field: FieldId, value: &Self::RustType) {
+        unsafe {
+            Schema_AddBool(object.internal, field, *value as u8);
+        }
+    }
+
+    fn get_field(object: &SchemaObject, field: FieldId) -> Self::RustType {
+        let raw = unsafe { Schema_GetBool(object.internal, field) };
+        raw != 0
+    }
+}
+
+impl IndexedField for bool {
+    fn field_count(object: &SchemaObject, field: FieldId) -> u32 {
+        unsafe { Schema_GetBoolCount(object.internal, field) }
+    }
+
+    fn index_field(object: &SchemaObject, field: FieldId, index: u32) -> Self::RustType {
+        let raw = unsafe { Schema_IndexBool(object.internal, field, index) };
+        raw != 0
+    }
+}
+
+impl<T: IndexedField> Sealed for Option<T> {}
+
+impl<T: IndexedField> SchemaField for Option<T> {
+    type RustType = Option<T::RustType>;
+
+    fn add_field<'a>(object: &mut SchemaObject<'a>, field: FieldId, value: &'a Self::RustType) {
+        if let Some(value) = value {
+            T::add_field(object, field, value);
+        }
+    }
+
+    fn get_field(object: &SchemaObject, field: FieldId) -> Self::RustType {
+        match T::field_count(object, field) {
+            0 => None,
+            _ => Some(T::get_field(object, field)),
+        }
+    }
+}
+
+impl<K, V> Sealed for BTreeMap<K, V>
+where
+    K: IndexedField,
+    V: IndexedField,
+    K::RustType: Ord,
+{
+}
+
+impl<K, V> SchemaField for BTreeMap<K, V>
+where
+    K: IndexedField,
+    V: IndexedField,
+    K::RustType: Ord,
+{
+    type RustType = BTreeMap<K::RustType, V::RustType>;
+
+    fn add_field(object: &mut SchemaObject, field: FieldId, map: &Self::RustType) {
+        // Create a schema object at the specified field in `object`.
+        let object: SchemaObject = unsafe { Schema_AddObject(object.internal, field) }.into();
```

Comment:
This still isn't quite right, you don't need this line. You just add key-value pair objects at the field id in the parent map.

Reply:
Similarly for the `get_field` implementation

Reply:
Oh, of course 🤦‍♀️ I'll get this right some day.